packages/langium/src/grammar/langium-grammar-validator.ts

```diff
@@ -269,13 +270,31 @@ export class LangiumGrammarValidator {
             const isInfers = !rule.returnType && !rule.dataType;
             const ruleTypeName = getTypeName(rule);
             if (!isDataType && ruleTypeName && types.has(ruleTypeName) === isInfers) {
-                const keywordNode = isInfers ? findNodeForKeyword(rule.$cstNode, 'infer') : findNodeForKeyword(rule.$cstNode, 'returns');
-                accept('error', getMessage(ruleTypeName, isInfers), {
-                    node: rule.inferredType ?? rule,
-                    property: 'name',
-                    code: isInfers ? IssueCodes.InvalidInfers : IssueCodes.InvalidReturns,
-                    data: keywordNode && toDocumentSegment(keywordNode)
-                });
+                if(rule.inferredType === undefined) {
```

Comment:
Suggestion: I believe there is quite a bit of overlap in these branches. This can be trimmed down by wrapping the `suggestion` property in a ternary.

Reply:
There is, I still want to go back over and rework this. Most of it is a slice from the pre-existing PR, and much has changed since then.

Reply:
Updated to consolidate